File: Sources/AgentToolCore/Migration.swift
import Foundation

public enum Migration {
    public enum Failure: Error, Equatable { case sourceMissing, conflict }

    public static func migrate(from source: URL, env: Environment) throws -> Int {
        let expected = env.home.appending(path: "Library/Application Support/ManageArms").standardizedFileURL
        guard source.standardizedFileURL == expected,
              FileManager.default.fileExists(atPath: source.appending(path: "registry.json").path) else {
            throw Failure.sourceMissing
        }
        let fm = FileManager.default
        let targets = [env.registryFile, env.agentStore, env.disabledStore, env.disabledAgentStore]
        guard !targets.contains(where: { fm.fileExists(atPath: $0.path) }) else { throw Failure.conflict }
        let legacy = Environment(home: env.home, appSupport: source, run: env.run, now: env.now, httpGet: env.httpGet)
        let registry = try Registry.read(env: legacy)
        let stage = URL(filePath: NSTemporaryDirectory()).appending(path: "agent-tool-migration-\(UUID().uuidString)")
        try fm.createDirectory(at: stage, withIntermediateDirectories: true)
        defer { try? fm.removeItem(at: stage) }
        var moved: [URL] = []
        var succeeded = false
        defer {
            if !succeeded {
                for target in moved where fm.fileExists(atPath: target.path) { try? fm.removeItem(at: target) }
            }
        }
        let copies = [("agents", env.agentStore), ("disabled-skills", env.disabledStore), ("disabled-agents", env.disabledAgentStore)]
        for (name, target) in copies where fm.fileExists(atPath: source.appending(path: name).path) {
            try fm.copyItem(at: source.appending(path: name), to: stage.appending(path: name))
            try fm.createDirectory(at: target.deletingLastPathComponent(), withIntermediateDirectories: true)
            try fm.moveItem(at: stage.appending(path: name), to: target)
            moved.append(target)
        }
        try registry.save(env: env)
        succeeded = true
        return registry.resources.count
    }
}

File: src/migration.ts
import { homedir } from "node:os";
import { join } from "node:path";
import { agentStore, disabledAgentStore, disabledStore, Env, registryFile } from "./env";
import { AgentToolError } from "./errors";
import { decode, save } from "./registry";
import * as guard from "./writeGuard";
import { readJsonc } from "./mcpScanner";

/**
 * 旧 ManageArms の保存領域。macOS 専用アプリなので、他の OS では検知そのものを行わない。
 */
export const legacyPath = (): string | null =>
  process.platform === "darwin"
    ? join(homedir(), "Library", "Application Support", "ManageArms")
    : null;

export const hasLegacyData = (): boolean => {
  const path = legacyPath();
  return path !== null && guard.exists(join(path, "registry.json"));
};

/**
 * 旧データを `globalStorageUri` へ移す。
 *
 * 移行先に既にデータがあれば上書きせず止める。旧ファイルには一切書き込まない —
 * 失敗しても元の環境はそのまま残り、次回起動で再試行できる。
 */
export async function migrate(source: string, env: Env):
  Promise<{ migratedEntries: number; skipped: number }> {
  const expected = legacyPath();
  if (expected === null || source !== expected || !guard.exists(join(source, "registry.json"))) {
    throw new AgentToolError("NOT_FOUND", "the ManageArms data directory was not found");
  }

  const stores = [
    ["agents", agentStore(env)],
    ["disabled-skills", disabledStore(env)],
    ["disabled-agents", disabledAgentStore(env)],
  ] as const;
  const targets = [registryFile(env), ...stores.map(([, target]) => target)];
  if (targets.some(guard.exists)) {
    throw new AgentToolError("MIGRATION_CONFLICT",
      "Agent Tool already holds data here; not overwriting it");
  }

  // browserDetection / menuBar / appearance は Mac App の設定。`decode` が落とす。
  const registry = decode(readJsonc(join(source, "registry.json")));

  // 検証してから移す。staging は OS の一時領域に置き、全経路で片付ける。
  const stage = guard.stagingDir("agent-tool-migration-");
  const moved: string[] = [];
  try {
    for (const [name, target] of stores) {
      const from = join(source, name);
      if (!guard.exists(from)) continue;
      guard.copy(from, join(stage, name));
      guard.prepare(target, join(target, ".."),
        guard.isInside(target, env.home) ? env.home : env.appSupport);
      guard.move(join(stage, name), target);
      moved.push(target);
    }
    await save(env, registry);
  } catch (error) {
    // 失敗したら今回作ったものだけを消す。source には触れない。
    for (const target of moved) {
      if (guard.exists(target)) guard.remove(target);
    }
    throw error;
  } finally {
    guard.remove(stage);
  }

  return {
    migratedEntries: registry.resources.length,
    skipped: stores.length - moved.length,
  };
}
